### PR TITLE
feat: adds support for multiple BLS keys

### PR DESCRIPTION
**Pull request by ckartik**
## Describe your changes

Added support for multiple BLS keys per provider in the `ProviderRegistry` and `BlockTracker` contracts. Updated functions to handle arrays of BLS keys, enhancing flexibility for provider registration and staking. Refactored the `recordL1Block` function to use BLS keys instead of graffiti strings to identify winning block builders. Also introduced a `setProviderRegistry` function in `BlockTracker` for dynamic registry assignments.

## Issue ticket number and link

Fixes #468

## Checklist before requesting a review

- [X] Added tests to validate multi-BLS key functionality for providers
- [X] Updated documentation to reflect changes in BLS key handling and new function

### CODE REVIEW

File: contracts/contracts/core/BlockTracker.sol
Comment (Mikelle) [inline contracts/contracts/core/BlockTracker.sol:95]: Would be great to generate an event here.

File: contracts/contracts/core/ProviderRegistry.sol
Comment (Mikelle) [inline contracts/contracts/core/ProviderRegistry.sol:314]: Maybe we could do that in the previous for loop, to avoid additional looping?

File: contracts/test/core/PreconfManagerTest.sol
Comment (Mikelle) [inline contracts/test/core/PreconfManagerTest.sol:41]: Would be great to add test with several pubkeys

File: oracle/pkg/l1Listener/l1Listener_test.go
Comment (Mikelle) [inline oracle/pkg/l1Listener/l1Listener_test.go:29]: why is it commented?

File: contracts/contracts/core/ProviderRegistry.sol
Comment (shaspitz) [inline contracts/contracts/core/ProviderRegistry.sol:323]: Keep newline

File: contracts/contracts/core/ProviderRegistryStorage.sol
Comment (shaspitz) [inline contracts/contracts/core/ProviderRegistryStorage.sol:47]: Keep newline

File: contracts/contracts/core/ProviderRegistryStorage.sol
Comment (shaspitz) [inline contracts/contracts/core/ProviderRegistryStorage.sol:40]:
```suggestion
   /// @dev Maps a BLS public key to its corresponding block builder address
```

File: contracts/contracts/interfaces/IProviderRegistry.sol
Comment (shaspitz) [inline contracts/contracts/interfaces/IProviderRegistry.sol:85]: Keep newline

File: infrastructure/nomad/playbooks/variables/profiles.yml
Comment (shaspitz) [inline infrastructure/nomad/playbooks/variables/profiles.yml:686]: why remove these?

File: contracts/contracts/core/BlockTracker.sol
Comment (Mikelle) [inline contracts/contracts/core/BlockTracker.sol:96]: Doesn’t this emit the event with the same two variables?

File: contracts/contracts/core/ProviderRegistry.sol
Comment (Mikelle) [inline contracts/contracts/core/ProviderRegistry.sol:285]:
```suggestion
    * @param blsPublicKeys BLS public keys of the provider.
```

File: contracts/contracts/core/ProviderRegistry.sol
Comment (Mikelle) [inline contracts/contracts/core/ProviderRegistry.sol:275]:
```suggestion
     * @param blsPublicKeys The BLS public keys of the provider.
```

File: contracts/contracts/core/ProviderRegistry.sol
Comment (Mikelle) [inline contracts/contracts/core/ProviderRegistry.sol:308]: nit: `!=` consumes a little bit less gas than `>`
```suggestion
        require(blsPublicKeys.length != 0, AtLeastOneBLSKeyRequired());
```

File: contracts/contracts/core/ProviderRegistry.sol
Comment (Mikelle) [inline contracts/contracts/core/ProviderRegistry.sol:310]: `bytes blsPublicKey = blsPublicKeys[i];`
Then use this variable. That way you will have less storage reads

File: contracts/test/core/BidderRegistryTest.sol
Comment (Mikelle) [inline contracts/test/core/BidderRegistryTest.sol:567]: nit: missing new line

File: contracts/test/core/BidderRegistryTest.sol
Comment (Mikelle) [inline contracts/test/core/BidderRegistryTest.sol:52]:
```suggestion
                (minStake, feeRecipient, feePercent, address(this), 24*hours, 5*hours)
```

File: contracts/test/core/OracleTest.sol
Comment (Mikelle) [inline contracts/test/core/OracleTest.sol:32]:
```suggestion
    uint256 public constant withdrawalDelay = 24 * hours; // 24 hours
```

File: contracts/test/core/OracleTest.sol
Comment (Mikelle) [inline contracts/test/core/OracleTest.sol:756]: nit: missing new line

File: contracts/test/core/PreconfManagerTest.sol
Comment (Mikelle) [inline contracts/test/core/PreconfManagerTest.sol:1174]: nit: missing new line

File: infrastructure/nomad/playbooks/templates/jobs/mev-commit-oracle.nomad.j2
Comment (Mikelle) [inline infrastructure/nomad/playbooks/templates/jobs/mev-commit-oracle.nomad.j2:186]: do we want to keep it like that?

File: infrastructure/nomad/playbooks/templates/jobs/mev-commit.nomad.j2
Comment (Mikelle) [inline infrastructure/nomad/playbooks/templates/jobs/mev-commit.nomad.j2:106]: do we want to keep it like that?

File: oracle/pkg/l1Listener/l1Listener.go
Comment (Mikelle) [inline oracle/pkg/l1Listener/l1Listener.go:190]:
```suggestion
				header, err := l.l1Client.HeaderByNumber(ctx, new(big.Int).SetUint64(b+1))
```

File: oracle/pkg/l1Listener/l1Listener.go
Comment (Mikelle) [inline oracle/pkg/l1Listener/l1Listener.go:206]: Why sleep is 500ms and not any other value?

File: oracle/pkg/l1Listener/l1Listener.go
Comment (Mikelle) [inline oracle/pkg/l1Listener/l1Listener.go:200]: why we are trying for 2 seconds and not for any other time?

File: p2p/pkg/preconfirmation/tracker/tracker.go
Comment (Mikelle) [inline p2p/pkg/preconfirmation/tracker/tracker.go:237]: Maybe it is better to keep it as a debug option?

File: p2p/pkg/preconfirmation/tracker/tracker.go
Comment (Mikelle) [inline p2p/pkg/preconfirmation/tracker/tracker.go:333]: Maybe it is better to keep it as a debug option?

File: oracle/pkg/l1Listener/l1Listener.go
Comment (ckartik) [reply oracle/pkg/l1Listener/l1Listener.go:200]: It seemed to work well, there's no strong reason.

Comment (ckartik) [reply oracle/pkg/l1Listener/l1Listener.go:200]: It's just to attempt around 4 retries.

File: oracle/pkg/l1Listener/l1Listener.go
Comment (ckartik) [reply oracle/pkg/l1Listener/l1Listener.go:205]: spending 2 seconds for 4 requests works well during my long run testing.

File: contracts/scripts/core/DeployCore.s.sol
Comment (shaspitz) [inline contracts/scripts/core/DeployCore.s.sol:35]: 24 hours right? If so we shouldn't multiply by 1000

File: contracts/contracts/core/PreconfManager.sol
Comment (shaspitz) [inline contracts/contracts/core/PreconfManager.sol:87]: Can remove this line, I changed it to `block.chanid` in other PR

File: contracts/scripts/core/DeployCore.s.sol
Comment (ckartik) [reply contracts/scripts/core/DeployCore.s.sol:35]: It's in miliseconds for our chain.

File: contracts/contracts/core/PreconfManager.sol
Comment (ckartik) [reply contracts/contracts/core/PreconfManager.sol]: Great!

File: cl/mockrelay/mock_relay.go
Comment (Mikelle) [inline cl/mockrelay/mock_relay.go:1]: why is this was added to `cl` package?

File: .github/workflows/infrastructure.yml
Comment (shaspitz) [inline .github/workflows/infrastructure.yml:218]: Intentionally committing the api key here?

File: contracts/scripts/core/DeployCore.s.sol
Comment (shaspitz) [inline contracts/scripts/core/DeployCore.s.sol:36]: A bit confusing to use `hours` here since this var is in a unit of blocks

File: contracts/test/core/BidderRegistryTest.sol
Comment (Mikelle) [inline contracts/test/core/BidderRegistryTest.sol:52]: nit: extra space

File: contracts/test/core/OracleTest.sol
Comment (Mikelle) [inline contracts/test/core/OracleTest.sol:32]: nit: extra space

File: oracle/pkg/l1Listener/l1Listener.go
Comment (shaspitz) [inline oracle/pkg/l1Listener/l1Listener.go:182]: What's the reasoning here? Not sure I understand the comment

File: oracle/pkg/l1Listener/l1Listener.go
Comment (Mikelle) [reply oracle/pkg/l1Listener/l1Listener.go:205]: It would be awesome to add a comment about that

File: oracle/pkg/l1Listener/l1Listener.go
Comment (Mikelle) [inline oracle/pkg/l1Listener/l1Listener.go:204]: In case of a error, is log supposed to be an info?

File: oracle/pkg/l1Listener/l1Listener.go
Comment (Mikelle) [inline oracle/pkg/l1Listener/l1Listener.go:209]: In case of a error, is log supposed to be an info?

**Review by shaspitz — COMMENTED** [review]
Looks good to me in general 👍, might be worth getting @aloknerurkar's stamp of approval on the oracle changes

File: contracts/contracts/core/BlockTracker.sol
Comment (shaspitz) [reply contracts/contracts/core/BlockTracker.sol:96]: Looks like this still needs to be fixed

File: oracle/pkg/l1Listener/l1Listener.go
Comment (ckartik) [reply oracle/pkg/l1Listener/l1Listener.go:204]: No this is the happy path, for blocks not going through PBS, they not show up on the relays.

File: oracle/pkg/l1Listener/l1Listener.go
Comment (ckartik) [reply oracle/pkg/l1Listener/l1Listener.go:209]: This is still a happy path operation, see my comment above.

**Review by shaspitz — APPROVED** [review]

**Comment by shaspitz** [comment]
Approved to unblock @ckartik, contract changes lgtm (after addressing final comments like event emitting etc). @aloknerurkar and @mrekucci would be useful to get your approval on the infra/oracle changes to confirm this doesn't break anything

File: oracle/pkg/l1Listener/l1Listener.go
Comment (aloknerurkar) [inline oracle/pkg/l1Listener/l1Listener.go:199]: Why not put the retry logic in the RelayQuerier itself? This unnecessarily complicates this loop.

File: oracle/pkg/l1Listener/l1Listener.go
Comment (aloknerurkar) [inline oracle/pkg/l1Listener/l1Listener.go:200]: Context should be passed to the Query function as this can take a long time.

File: oracle/pkg/l1Listener/l1Listener.go
Comment (aloknerurkar) [inline oracle/pkg/l1Listener/l1Listener.go:254]: Why mini?

File: oracle/pkg/l1Listener/l1Listener.go
Comment (aloknerurkar) [inline oracle/pkg/l1Listener/l1Listener.go:267]: This function should be restructured ideally. The top context should be passed here and you can include the retry logic here itself to simplify the l1 listener loop.

Secondly, if you only care about the first result, you should use errgroup. With errgroup you will get a context cancellation when the first routine exits. You can use this context in the http.RequestWithContext function so that the other requests would be cancelled on the first result.

For the retry logic you can use the backoff library that we added in the new consensus changes (https://github.com/primev/mev-commit/blob/main/cl/redisapp/util/retry.go#L25). Each go routine can retry with backoff till you get the first result.

File: p2p/rpc/providerapi/v1/providerapi.proto
Comment (aloknerurkar) [inline p2p/rpc/providerapi/v1/providerapi.proto:95]: We should also add check for len > 0

**Review by aloknerurkar — CHANGES_REQUESTED** [review]

File: .github/workflows/infrastructure.yml
Comment (aloknerurkar) [inline .github/workflows/infrastructure.yml:218]: This should be removed

File: oracle/pkg/l1Listener/l1Listener.go
Comment (mrekucci) [inline oracle/pkg/l1Listener/l1Listener.go:264]: Use `url.PathEscape` for the path.

File: oracle/pkg/l1Listener/l1Listener.go
Comment (mrekucci) [inline oracle/pkg/l1Listener/l1Listener.go:300]: `fmt.Sprintf("%d", blockNumber)` use `strconv.Atoi()` instead.

File: oracle/pkg/l1Listener/l1Listener.go
Comment (mrekucci) [inline oracle/pkg/l1Listener/l1Listener.go:301]: What is the expected value here?

File: p2p/integrationtest/provider/client.go
Comment (mrekucci) [inline p2p/integrationtest/provider/client.go:114]: The context should have a timeout.

File: oracle/pkg/l1Listener/l1Listener.go
Comment (ckartik) [reply oracle/pkg/l1Listener/l1Listener.go:319]: The builders public key.

**Review by aloknerurkar — APPROVED** [review]

**Pull request merged**
